pkg/controllers/workapplier/pq.go

```
@@ -0,0 +1,229 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	fleetv1beta1 "github.com/kubefleet-dev/kubefleet/apis/placement/v1beta1"
+)
+
+const (
+	// A list of priority levels and their targets for the work applier priority queue.
+	//
+	// The work applier, when a priority queue is in use, will prioritize requests in the following
+	// order:
+	// * with highest priority (-2): all Create/Delete events, and all Update events
+	//   that concern recently created Work objects or Work objects that are in a failed/undeterminted
+	//   state (apply op/availability check failure, or diff reporting failure).
+	// * with medium priority (-1): all other Update events.
+	// * with default priority (0): all requeues (with or with errors), and all Generic events.
+	//
+	// Note that requests with the same priority level will be processed in the FIFO order.
+	//
+	// TO-DO (chenyu1): evaluate if/how we need to/should prioritize requeues properly.
+	highPriorityLevel    = 2
+	mediumPriorityLevel  = 1
+	defaultPriorityLevel = 0
+)
+
+type CustomPriorityQueueManager interface {
+	PriorityQueue() priorityqueue.PriorityQueue[reconcile.Request]
+}
+
+var _ handler.TypedEventHandler[client.Object, reconcile.Request] = &priorityBasedWorkObjEventHandler{}
+
+// priorityBasedWorkObjEventHandler implements the TypedEventHandler interface.
+//
+// It is used to process work object events in a priority-based manner with a priority queue.
+type priorityBasedWorkObjEventHandler struct {
+	qm                                 CustomPriorityQueueManager
+	workObjAgeForPrioritizedProcessing time.Duration
+}
+
+// Create implements the TypedEventHandler interface.
+func (h *priorityBasedWorkObjEventHandler) Create(_ context.Context, createEvent event.TypedCreateEvent[client.Object], _ workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	// Do a sanity check.
+	//
+	// Normally when this method is called, the priority queue has been initialized.
```

Why would the priority queue not be initialized when this method is called?

---

yeah but I guess the "out of our control" argument is valid. Maybe just comment on exactly why you need the sanity check - because it is out of our control while admitting the risk of silently dropping events?